### PR TITLE
Allow finer-grained validation suppression

### LINKED ISSUE #719: Finer grained validation suppression
The validation package allows turning off specific validation checks for specific files, including the use of a wildcard for either check or file.  It might be easier in the future to have a finer-grained specification for disabling checks, such as by date, geo resolution, and/or signal.  In particular the date wildcard seems important since the validation period will likely be set to be *n* days in the past until today; to block a check on a finer granularity than all files, you would have to edit the list of validations to suppress every day.

Tagging @nmdefries because he has probably thought more about what sorts of use cases we would actually have for different granularities.

### PR DESCRIPTION
**Pull request by sgsmob**
### Description
Change the way we disable specific validation checks to allow disabling based on  geo resolution, signal, and date.

To accommodate these additional parameters, the `params.json` syntax for disabling checks has changed from a list of length-2 lists to a list of objects whose keys are some subset of `check_name`, `geo_type`, `date`, and `signal`.  Any of these parameters absent from the object are interpreted as wildcards (i.e. `check_name: a, geo_type: b` disables `check_name: a, geo_type:b, signal:c, date: 2021-01-24`.

### Changelog
- `errors.py`:  overhaul to `ValidationFailure`
    - `None` is now the wildcard
    - Use custom `__eq__()` method for comparison
- `validate.py`:  change method of reading suppressions from `params.json`
- `datafetcher.py`, `dynamic.py`, `static.py`:  change `ValidationFailure` constructor
- `report.py`:  pass around `ValidationFailure` objects in suppression list

### Fixes 
- Fixes #719

### CODE REVIEW

**Review by melange396 — APPROVED** [review]
looks pretty good to me.  i have just one real point of confusion in `params.json.template`

File: validator/README.md [hunk 1]
```diff
@@ -55,7 +55,11 @@ Please update the follow settings:
...
    * `end_date`: specifies the last date to be checked; if set to "latest", `end_date` will always be the current date
    * `span_length`: specifies the number of days before the `end_date` to check. `span_length` should be long enough to contain all recent source data that is still in the process of being updated (i.e. in the backfill period), for example, if the data source of interest has a 2-week lag before all reports are in for a given date, `scan_length` should be 14 days
-   * `suppressed_errors`: list of pairs of (`check_name`, `file_name`) uniquely specifying errors that have been manually verified as false positives or acceptable deviations from expected.  Either value can also take on the value `*` to apply to all check or file names.
+   * `suppressed_errors`: list of objects specifying errors that have been manually verified as false positives or acceptable deviations from expected.  These errors can be specified with the following variables, where omitted values are interpretted as a wildcard, i.e., not specifying a date applies to all dates:
```
[the remark this answers is not in the document]
Comment (sgsmob) [reply validator/README.md]: Done

File: validator/delphi_validator/errors.py [hunk 2]
```diff
@@ -2,7 +2,9 @@
 """
 Custom validator exceptions.
 """
-from dataclasses import dataclass
+# from dataclasses import dataclass
```
[the remark this answers is not in the document]
Comment (sgsmob) [reply validator/delphi_validator/errors.py]: Done

File: validator/params.json.template [hunk 3]
```diff
@@ -5,8 +5,13 @@
...
       "suppressed_errors": [
-        ["check_min_max_date", "county", "confirmed_7dav_cumulative_prop"],
-        ["check_val_lt_0", "20200906_county_deaths_7dav_incidence_num.csv"]],
+        {"check_name": "check_min_max_date",
+         "geo_type": "county",
+         "signal": "confirmed_7dav_cumulative_prop"},
+        {"check_name": "check_val_lt_0",
+         "date": "2020-09-06",
+         "geo_type": "county",
+         "signal": "deaths_7dav_incidence_num.csv"}],
```
[the remark this answers is not in the document]
Comment (sgsmob) [reply validator/params.json.template]: The `.csv` extension was a mistake here.  Fixed

**Comment by krivard** [comment]
Is this ready to merge?

**Pull request merged**